subworkflows/local/metagenomics_profiling.nf

```diff
@@ -0,0 +1,160 @@
+
+workflow METAGENOMICS_PROFILING {
+
+    take:
+    reads
+    database
+
+    main:
+
+    ch_versions                  = Channel.empty()
+    ch_raw_classifications       = Channel.empty()
+    ch_raw_profiles              = Channel.empty()
+    ch_multiqc_files             = Channel.empty()
+    ch_postprocessing_input      = Channel.empty()
+
+    /*
+        PREPARE PROFILER INPUT CHANNELS & RUN PROFILING
+    */
+
+    // Each tool has a slightly different input structure and generally separate
+    // input channels for reads vs database. We restructure the channel tuple
+    // for each tool and make liberal use of multiMap to keep reads/database
+    // channel element order in sync with each other
+
+    if ( params.metagenomics_profiling_tool == 'malt' ) {
+
+        // Reset entire input meta for MALT to just database name,
+        // as we don't run run on a per-sample basis due to huge databases
+        // so all samples are in one run and so sample-specific metadata
+        // unnecessary. Set as database name to prevent `null` job ID and prefix.
+
+        if ( params.metagenomics_malt_group_size > 0 ) {
+            ch_input_for_malt_tmp =  reads
+                .map { meta, reads -> reads }
+                .collate( params.metagenomics_malt_group_size ) //collate into bins of defined lengths
+                .map{
+                    reads ->
+                    // add new meta with db-name as id
+                    [[id: file(params.metagenomics_profiling_database).getBaseName() ], reads]
+                }
+                .combine(database) //combine with database
+                .multiMap{
+                    // and split apart again
+                    meta, reads, database ->
+                        reads: [meta, reads]
+                        database: database
+                }
+            ch_input_for_malt = ch_input_for_malt_tmp.reads
+            database = ch_input_for_malt_tmp.database
+        }
+
+        else {
+            ch_input_for_malt =  reads
+                .map { meta, reads -> reads }
+                .collect()
+                .map{
+                    // make sure id is db_name for publishing purposes.
+                    reads ->
+                    [[id: file(params.metagenomics_profiling_database).getBaseName() ], reads]
+                }
+        }
+
+        // Run MALT
+        MALT_RUN ( ch_input_for_malt, database )
+
+        ch_maltrun_for_megan = MALT_RUN.out.rma6
+            .transpose()
+            .map {
+                meta, rma ->
+                    // re-extract meta from file names, use filename without rma to
+                    // ensure we keep paired-end information in downstream filenames
+                    // when no pair-merging
+                    def meta_new = meta.clone()
+                    meta_new['db_name'] = meta.id
+                    meta_new['id'] = rma.baseName
+                [ meta_new, rma ]
```

Comment:
We don't do clones anymore, https://www.youtube.com/watch?v=A357C-ux6Dw&t=507s&pp=ygURcm9iIHN5bWUgbmV4dGZsb3c%3D

Or see taxprofiler or mag for examples :)

Reply:
back to good old "+"? :D